prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py

```diff
@@ -273,3 +269,104 @@ def get_efs_vmt_projection_for_state(efs_annual_data, electrification_scenario=N
     }
 
     return efs_for_state
+
+
+def generate_scaling_factor(efs_vmt_for_state, vmt_for_ua_perc, vmt_for_ra_perc):
+    """Calculate for each year the scaling factors by vehicle technology in uraban and
+    rural areas.
+
+    :param dict efs_vmt_for_state: keys are state abbreviations and values are data
+        frames enclosing the projected annual annual VMT by vehicle technology and year
+        as returned by :func:`get_efs_vmt_projection_for_state`.
+    :param dict vmt_for_ua_perc: keys are state abbreviations and values series giving
+        VMT for urban areas in state. THis is returned by
+        :func:`calculate_urban_rural_fraction`
+    :param dict vmt_for_ua_perc: same as ``vmt_for_ua_perc`` but for rural areas.
+    :return: (*tuple*) -- dictionary for uraban and rural area. Keys are years, values
+        are data frames enclosing scaling factors for each vehicle type.
+    """
+    old2new = {
+        "ELECTRIC LIGHT-DUTY AUTO - 100 MILE RANGE": "LDV Car - 100 mi",
+        "ELECTRIC LIGHT-DUTY AUTO - 200 MILE RANGE": "LDV Car - 200 mi",
+        "ELECTRIC LIGHT-DUTY AUTO - 300 MILE RANGE": "LDV Car - 300 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 100 MILE RANGE": "LDV Truck - 100 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 200 MILE RANGE": "LDV Truck - 200 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 300 MILE RANGE": "LDV Truck - 300 mi",
+        "BATTERY ELECTRIC MEDIUM-DUTY VEHICLE": "MDV Truck",
+        "ELECTRIC HEAVY DUTY VEHICLE": "HDV Truck",
+    }
+
+    rural_scaling_factor = {}
+    urban_scaling_factor = {}
+    for s in state2abv.values():
+        efs_value = efs_vmt_for_state[s].pivot(
+            index="YEAR", columns="DEMAND_TECHNOLOGY", values="VALUE"
+        )
+        for y in efs_value.index:
+            rural_scaling_factor[y] = pd.concat(
+                [
+                    rural_scaling_factor.get(y, pd.DataFrame()),
+                    pd.DataFrame(
+                        (efs_value.loc[y] * vmt_for_ra_perc[s].values).to_dict(),
```

Comment:
Not sure whether we can do this multiplication outside the loop given it is independent of year. Same comment below for the UA calculations.

Reply:
IDone for rural. For urban we would to create a multi index data frame (year and urban areas) and it makes things more complicated.